include/libtorrent/aux_/utp_stream.hpp

```diff
@@ -971,6 +973,10 @@ struct utp_socket_impl
 	// the socket being writable again
 	bool m_stalled:1;
 
+	// if the socket was 'stalled' when we wanted to send the ST_FIN packet
+	// then we retry once the socket is writable again
+	bool m_stalled_fin:1;
```

Comment:
it seems reasonable to keep this state, but I wonder if it's not possible to infer it from `state() == state_t::fin_sent`.

We the `fin_sent` state means we're not going to transmit any more (new) payload, but possibly resend lost packets. I suppose we can't necessarily know that if the socket is un-stalled, that we need to resend the FIN packet just because we're in the fin_sent state.

Reply:
I figure that if we reach `writable()` while in the fin-sent state, we can't really tell if we've actually sent the fin packet or not - we could be there after a stall from a resend or ack. I figured explicit noting whether we need to call send_fin() again would be the simplest way.

Although, maybe we could check `m_sendbuf` for the sent-but-unack'd fin packet instead?

Reply:
`(state() == state_t::fin_sent && m_outbuf.at(m_seq_nr) == nullptr)`

Would this be able to pick it up? Once we're in the fin_sent state, `m_seq_nr` is the fin packet itself. I think the only time it won't be present in `m_outbuf` while in the fin_sent state would be if the `send_pkt(pkt_fin)` within `send_fin()` happened while stalled. Once it gets ack'd and removed from `m_outbuf`, the state progresses beyond fin_sent.